server/route.go

```diff
@@ -514,13 +520,75 @@ func (c *client) processRouteInfo(info *Info) {
+			// Update the name marking that no longer susceptible to dynamic cluster updates since
+			// there is an explicit name in the cluster and cause other nodes to reconnect to
+			// this node using the explicit name.
+			s.setClusterName(info.Cluster, false)
 			s.removeAllRoutesExcept(c)
 			c.mu.Lock()
+
+			c.route.clusterName = info.Cluster
+			c.route.clusterDynamic = info.Dynamic
+		} else if !isDynamic && remoteIsDynamic {
+			c.srv.mu.Lock()
+			routeInfoJSON := s.routeInfoJSON
+			c.srv.mu.Unlock()
+
+			c.mu.Lock()
+
+			// In case there is a mismatch with what we know about this route,
+			// we will send it another INFO message with the explicit cluster name
+			// so that the remote dynamic node uses our cluster name.
+			if c.route.clusterName != info.Cluster {
+				c.enqueueProto(routeInfoJSON)
+			}
+
+			// Override the cluster name from this route with ours since it should win.
+			c.route.clusterName = s.info.Cluster
+			c.route.clusterDynamic = info.Dynamic
+		} else if isDynamic && remoteIsDynamic {
+			if isSusceptible {
+				// Ignore the message unless we are susceptible to dynamic cluster updates and their
+				// dynamic cluster name is a better fit than what we have. If we are not susceptible,
+				// it means that there is an explicit name in the cluster and eventually nodes with the
+				// explicit cluster name will make this node use the explicit cluster name via INFO messages.
+				if strings.Compare(clusterName, info.Cluster) < 0 {
+					s.setClusterName(info.Cluster, true)
+					s.removeAllRoutesExcept(c)
+				}
+			} else if !s.nonDynamicClusterNameRoutePresent(info.ID) {
+				// If the route moved from being static to dynamic (thus also changing its name),
+				// then only use their name if no other static, non dynamic nodes exist and it beats
+				// this node's default unique cluster name.  This can happen for example when the last node
+				// from a cluster of nodes has its config reloaded to remove the name and go back to dynamic mode.
+
+				// Find any other route with an explicit name but the one that just went back to dynamic mode.
+				// NOTE: Sometimes nodes with dynamic names can also send their INFOs but due to races,
+				// by the time others have received it they have already changed its name to an explicit one
+				// due to a CONNECT message that was sent.  This check also covers that scenario.
+
+				// Use the correct name and recluster with that name.
+				if strings.Compare(s.defaultClusterName, info.Cluster) < 0 {
+					// Use theirs, let other nodes reconnect using their cluster name.
+					s.setClusterName(info.Cluster, true)
+					s.removeAllRoutesExcept(c)
+				} else {
+					// Use ours, let other nodes reconnect using our cluster name.
+					s.setClusterName(s.defaultClusterName, true)
+					s.removeAllRoutesExcept(c)
+				}
+			}
+
+			c.mu.Lock()
+			c.route.clusterName = info.Cluster
+			c.route.clusterDynamic = info.Dynamic
```

Comment:
May need to look at `addRoute()` when a route already existed but we "upgrade" to solicited, we copy over info, check `if r != nil {..`. Not saying that we need to put that there, just think if it applies.